### PR TITLE
Add support for parsing multiple types of quotation marks in commands, Fix #942

### LINKED ISSUE #942: Quotation marks characters
People that are using Discord on iPhone don't have classic quotation marks (Found out that it changed default's quotes after upgrade to iOS 11, its called smart quotes). Is it possible to extend CommandParser to handle these quotes?

For comparison:
Classic quotes - "text" [1](https://unicode-table.com/en/0022/)
Apple's smart quotes - “text” [2](https://unicode-table.com/en/201C/) [3](https://unicode-table.com/en/201D/)

### PR DESCRIPTION
**Pull request by Chris-Johnston**
#942 brings up the issue that non-remainder string parameters of commands cannot be parsed with non-standard double quotation marks (`"`). I added a list of quotation mark aliases to the `CommandServiceConfig`. This means the library can now parse unicode double quotation marks and also culture-specific ones (like `「` and `«`).

I used aliases that were mentioned in the d.net channel, but I'm not certain what the default values should be set to.

I would like to get feedback if anything needs improvement. My discord handle in the channel is `@ChrisJ#8703`.

**Testing Method:**
I wanted to test that a single string could be parsed correctly when wrapped in multiple types of quotation marks. Here's a command that I used:
```cs
[Command("testcmd")]
public async Task TestCmd(string txt, int aNumber)
{
    await ReplyAsync($"{txt} {aNumber}");
}
```

And sample execution yields:
`-testcmd "test" 123` => `test 123` (normal behavior)
`-testcmd “test\“cmd\"“ 123` => `test“cmd" 123` (escaping still works)
`-testcmd «this is a multiple word string» 123` => `this is a multiple word string 123`
`-testcmd «you can mix and match" 456` => `you can mix and match 456`

### CODE REVIEW

**Comment by Emzi0767** [comment]
I don't think mix and match is correct behaviour. What happens if: `!test «test" argument»`? Your code would make it throw or parse otherwise incorrectly, whereas it should pass `test" arguments`.

For reference, here's a map of various quote kinds:
- `"` -> `"`
- `'` -> `'`
- `«` -> `»`
- `‘` -> `’`
- `“` -> `”`
- `„` -> `‟`
- `‹` -> `›`
- `‚` -> `‛`
- `《` -> `》`
- `〈` -> `〉`
- `「` -> `」`
- `『` -> `』`
- `〝` -> `〞`
- `﹁` -> `﹂`
- `﹃` -> `﹄`
- `＂` -> `＂`
- `＇` -> `＇`
- `｢` -> `｣`

**Comment by Chris-Johnston** [comment]
For the purposes of outlining how I would approach fixing mismatched quotation marks, here's a gist that shows how I would do it it: https://gist.github.com/Chris-Johnston/37b37e70246562c27f0a42d6bd8eda74

Using this, I've been able to produce the correct output.

image

Do you think that this is the correct approach?

**Comment by Emzi0767** [comment]
I think you should not allow mix and match behaviour at all. Just match the outermost quote.

**Comment by khionu** [comment]
Mmm. Mix and match is an anti-pattern. Additionally, one might use one type of quotes in order to allow regular quotes inside the used quotes

**Comment by Chris-Johnston** [comment]
Changing this to a map was a good idea.
Now using the same command I'm able to do: `-testcmd «all ‘the “quotes 「」”’  ! ! " " » 123` and get back `all ‘the “quotes 「」”’  ! ! " "  123`. 

So it's allowing for different types of quotation marks to be nested without needing to be escaped.
Mix and match is no longer allowed.

File: src/Discord.Net.Commands/CommandService.cs [hunk 1]
```diff
@@ -45,6 +46,7 @@ public CommandService(CommandServiceConfig config)
             _ignoreExtraArgs = config.IgnoreExtraArgs;
             _separatorChar = config.SeparatorChar;
             _defaultRunMode = config.DefaultRunMode;
+            _quotationMarkAliasMap = config.QuotationMarkAliasMap;
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/CommandService.cs]: Make an `ImmutableDictionary` out of it here?

Well, I guess the consideration is: do we want to guarantee immutable semantics internally, or fake it and take a perf win but also allow the map to be mutated while the client is running (which isn't too hard to do while implemented this way)? cc: @foxbot @FiniteReality 

Either way, I'd make the storage field an `IReadOnlyDictionary`.

File: src/Discord.Net.Commands/CommandParser.cs [hunk 2]
```diff
@@ -13,7 +14,33 @@ private enum ParserPart
             Parameter,
             QuotedParameter
         }
-        
+
+        private static bool isOpenQuote(Dictionary<char,char> map, char c)
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/CommandParser.cs]: * Use PascalCase
* Can take the map in as an `IReadOnlyDictionary`
* Can be made a local function inside `ParseArgsAsync`

(Same goes for `getMatch`)

**Review by Joe4evr — CHANGES_REQUESTED** [review]
Can do with a few minor tweaks.

File: src/Discord.Net.Commands/CommandService.cs (hunk 1 above)
Comment (Joe4evr) [reply src/Discord.Net.Commands/CommandService.cs]: For completion's sake: There *is* a third option where we can fake the immutable semantics, but make it much harder for users to modify the map after initialization. (Still wouldn't *completely* prevent it, though.)

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandService.cs]: For now, I've left this as an `IReadOnlyDictionary`. I found that when the dictionary could be modified by the user externally it didn't break anything, but I could understand that being a poor design decision.

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandService.cs]: Oh, I didn't see your second comment. I think I ended up doing your third option. I've been able to test it like this:

```cs
CommandServiceConfig config = new CommandServiceConfig();
var d = new Dictionary<char, char>()
{
    { '-', '-' },
    {'\"', '\"' }
};
config.QuotationMarkAliasMap = d;
commands = new CommandService(config);
d.Add('~', '~');
```

Again, I've found that it still works but might not be the intended idea.

Comment (Joe4evr) [reply src/Discord.Net.Commands/CommandService.cs]: Nah, that's the first thing that I talked about. The third option would be to do
```cs
_quotationMarkAliasMap = new Dictionary(config.QuotationMarkAliasMap);
```

Also, the type on the `CommandServiceConfig` *didn't* need to be changed to `IReadOnlyDictionary`. This now prohibits doing the following:
```cs
new CommandServiceConfig()
{
    QuotationMarkAliasMap =
    {
        { '[', ']' } //for example
    }
};
```

Comment (Joe4evr) [reply src/Discord.Net.Commands/CommandService.cs]: But anyway, I'd like to hear fox and finite's inputs. Do we want the internals to be semantically correct, or not take a perf hit?

File: src/Discord.Net.Commands/CommandParser.cs [hunk 3]
```diff
@@ -13,7 +14,7 @@ private enum ParserPart
             Parameter,
             QuotedParameter
         }
-        
+
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandParser.cs]: Please remove unnecessary whitespace

File: src/Discord.Net.Commands/CommandParser.cs [hunk 4]
```diff
@@ -24,7 +25,32 @@ private enum ParserPart
...
+                return c == '\"';
+            }
+
+            char GetMatch(IReadOnlyDictionary<char, char> dict, char ch)
+            {
+                // get the corresponding value for the key, if it exists
+                if (dict != null)
+                {
+                    char value;
+                    if (dict.TryGetValue(c, out value))
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandParser.cs]: `out var value` would probably preferable here

File: src/Discord.Net.Commands/CommandParser.cs [hunk 5]
```diff
@@ -97,7 +125,7 @@ private enum ParserPart
                 }
                 else if (curPart == ParserPart.QuotedParameter)
                 {
-                    if (c == '\"')
+                    if(c == matchQuote)
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandParser.cs]: Missing whitespace after if

File: src/Discord.Net.Commands/CommandParser.cs [hunk 6]
```diff
@@ -74,9 +100,11 @@ private enum ParserPart
                             argBuilder.Append(c);
                             continue;
                         }
-                        if (c == '\"')
+                        
+                        if(IsOpenQuote(command._quotationAliases, c))
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandParser.cs]: Missing whitespace after if

File: src/Discord.Net.Commands/CommandService.cs (hunk 1 above)
Comment (FiniteReality) [reply src/Discord.Net.Commands/CommandService.cs]: I'd say make an ImmutableDictionary out of it, because that's what we do in other places .

File: src/Discord.Net.Commands/CommandServiceConfig.cs [hunk 7]
```diff
@@ -16,6 +17,30 @@ public class CommandServiceConfig
         /// <summary> Determines whether RunMode.Sync commands should push exceptions up to the caller. </summary>
         public bool ThrowOnError { get; set; } = true;
 
+        /// <summary> Collection of aliases that can wrap strings for command parsing. 
+        /// represents the opening quotation mark and the value is the corresponding closing mark.</summary>
+        public Dictionary<char, char> QuotationMarkAliasMap { get; set; }
+        = new Dictionary<char, char>()
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandServiceConfig.cs]: Redundant parentheses here. Do we need the type name too?

File: src/Discord.Net.Commands/Info/CommandInfo.cs [hunk 8]
```diff
@@ -20,6 +20,7 @@ public class CommandInfo
 
         private readonly CommandService _commandService;
         private readonly Func<ICommandContext, object[], IServiceProvider, CommandInfo, Task> _action;
+        internal readonly IReadOnlyDictionary<char,char> _quotationAliases;
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/Info/CommandInfo.cs]: IMO this should not be stored on the command, but I'm not sure if we have the ability to pass the command service to the parser where ParseArgsAsync is called. Might be worth taking a look?

**Review by FiniteReality — CHANGES_REQUESTED** [review]

File: src/Discord.Net.Commands/CommandServiceConfig.cs (hunk 7 above)
Comment (Joe4evr) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: Type name is required until dotnet/csharplang#100 lands. Do we omit the parentheses in other places too?

Comment (FiniteReality) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: I'm pretty sure we omit the parentheses where possible.

File: src/Discord.Net.Commands/Info/CommandInfo.cs (hunk 8 above)
Comment (Chris-Johnston) [reply src/Discord.Net.Commands/Info/CommandInfo.cs]: I looked at this again, and found that I could use the similar behavior of `IgnoreExtraArgs`.

https://github.com/Chris-Johnston/Discord.Net/blob/46e9cf64f3d3635de872720c308e922b13ab8114/src/Discord.Net.Commands/Info/CommandInfo.cs#L124

File: src/Discord.Net.Commands/CommandServiceConfig.cs (hunk 7 above)
Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: GitHub doesn't seem to show this as outdated correctly, but please see commit 8187f35975031971b17818aae516fdfaf5d2eb36

**Comment by Joe4evr** [comment]
LGTM :shipit:

File: src/Discord.Net.Commands/CommandServiceConfig.cs [hunk 9]
```diff
@@ -1,4 +1,5 @@
-﻿namespace Discord.Commands
+﻿using System.Collections.Generic;
+namespace Discord.Commands
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandServiceConfig.cs]: Newline between these for consistency

**Review by FiniteReality — CHANGES_REQUESTED** [review]
Just a few more niggling things, and it'll be ready to merge

File: src/Discord.Net.Commands/CommandParser.cs [hunk 10]
```diff
@@ -12,9 +13,8 @@ private enum ParserPart
             None,
             Parameter,
             QuotedParameter
-        }
-        
-        public static async Task<ParseResult> ParseArgsAsync(CommandInfo command, ICommandContext context, bool ignoreExtraArgs, IServiceProvider services, string input, int startPos)
+        }
```
Comment (SubZero0) [inline src/Discord.Net.Commands/CommandParser.cs]: Missing new line between enum and method

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandParser.cs]: I must have mistook "remove unnecessary whitespace" as the line instead of the space after. Fixed now.

**Review by FiniteReality — APPROVED** [review]

File: src/Discord.Net.Commands/CommandParser.cs [hunk 11]
```diff
@@ -24,7 +25,31 @@ private enum ParserPart
...
+                if (dict != null)
+                    return dict.ContainsKey(ch);
+                // or otherwise if it is the default double quote
+                return c == '\"';
+            }
+
+            char GetMatch(IReadOnlyDictionary<char, char> dict, char ch)
+            {
+                // get the corresponding value for the key, if it exists
+                if (dict != null)
```
Comment (AntiTcb) [inline src/Discord.Net.Commands/CommandParser.cs]: This could just be reduced down to `(dict != null && dict.TryGetValue(c, out var value))`.

**Comment by ufcpp** [comment]
Here is open/close and initial/final quote punctuation pairs generated by https://gist.github.com/ufcpp/5b2cf9a9bf7d0b8743714a0b88f7edc5

```
()༺༻༼༽᚛᚜‘’“”‹›⁅⁆⌈⌉⌊⌋❨❩❪❫❬❭❮❯❰❱❲❳❴❵⟅⟆⟦⟧⟨⟩⟪⟫⟬⟭⟮⟯⦃⦄⦅⦆⦇⦈⦉⦊⦋⦌⦍⦎⦏⦐⦑⦒⦓⦔⦕⦖⦗⦘⧘⧙⧚⧛⧼⧽⸂⸃⸄⸅⸉⸊⸌⸍⸜⸝⸠⸡⸢⸣⸤⸥⸦⸧⸨⸩〈〉《》「」『』【】〔〕〖〗〘〙〚〛〝〞
```

File: src/Discord.Net.Commands/CommandServiceConfig.cs [hunk 12]
```diff
@@ -38,7 +38,56 @@ public class CommandServiceConfig
...
+            {'⸠', '⸡' },
+            {'⸢', '⸣' },
+            {'⸤', '⸥' },
+            {'⸦', '⸧' },
+            {'⸨', '⸩' },
+            {'【', '】'},
+            {'〔', '〕' },
+            {'〖', '〗' },
+            {'〘', '〙' },
+            {'〚', '〛' }
```
Comment (FiniteReality) [inline src/Discord.Net.Commands/CommandServiceConfig.cs]: IMO if we're adding a large dict of things like this, it might be better to use a separate file for it instead. I'm not sure though.

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: Do you mean deserializing the dictionary from another file, or another file in the source that defines the default value of this dictionary?

I think that including all of these by default is a good idea. Alternatively we could shorten this list and include a more complete one in the documentation.

Comment (FiniteReality) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: Another file in source, yeah, like the emoji map text template. It may be possible for us to re-use some of the code suggested by @ufcpp so that we can generate these automatically and keep them up-to-date when Unicode inevitably changes.

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandServiceConfig.cs]: It looks like that gist uses features that aren't supported in .netstandard1.1 (or 1.3) https://docs.microsoft.com/en-us/dotnet/api/system.char.getunicodecategory?view=netcore-2.0#Applies_to 
https://docs.microsoft.com/en-us/dotnet/api/system.string.normalize?view=netcore-2.0#Applies_to

So for now it may be best to define this dictionary in another file and leave it, unless/until we target netstandard2

**Comment by Chris-Johnston** [comment]
#1043 ~~~replaces ns1.3 with ns2.0~~~ targets netstandard2.0. if it is merged, we will be able to support generating quotation mark pairs programmatically, see the comment thread immediately above

Edit: I misread. Ns2.0 is added here, but not replacing ns1.3. So I'm not sure if the generation would be viable.

File: src/Discord.Net.Commands/CommandParser.cs (hunk 11 above)
Comment (acid-chicken) [reply src/Discord.Net.Commands/CommandParser.cs]: @AntiTcb How about `(dict?.TryGetValue(c ,out var value) ?? false)`?

Comment (Joe4evr) [reply src/Discord.Net.Commands/CommandParser.cs]: Or just make sure that the collection that gets passed in is always initialized so that it's at least empty?

Comment (Chris-Johnston) [reply src/Discord.Net.Commands/CommandParser.cs]: I like the idea of assuming that the collection is always initialized. Added in 820f9e36483d63df64ae651f862d0226c4c8e52a
